python-sdk/nuscenes/predict/helper.py

```diff
@@ -0,0 +1,181 @@
+    :param translation: Tuple of (x, y, z) location that is the center of the new frame
+    :param rotation: Tuple representation of quaternion of new frame."""
+
+    yaw = angle_of_rotation(quaternion_yaw(Quaternion(rotation)))
+
+    transform = np.array([[np.cos(yaw), -np.sin(yaw)], [np.sin(yaw), np.cos(yaw)]])
+
+    coords = (coordinates - np.atleast_2d(np.array(translation)[:2])).T
+
+    return np.dot(transform, coords).T[:, :2]
+
+class PredictHelper:
+    """Wrapper class around NuScenes to help retrieve data for the prediction task."""
+
+    def __init__(self, nusc: NuScenes):
+        """Inits PredictHelper
+        :param nusc: Instance of NuScenes class
+        """
+        self.data = nusc
+        self.inst_sample_to_ann = self._map_sample_and_instance_to_annotation()
+
+    def _map_sample_and_instance_to_annotation(self) -> Dict[Tuple[str, str], str]:
+        """Creates mapping to look up an annotation given a sample and instance in constant time."""
+        mapping = {}
+
+        for record in self.data.sample_annotation:
+            mapping[(record['sample_token'], record['instance_token'])] = record['token']
+
+        return mapping
+
+    def _timestamp_for_sample(self, sample: str):
+        """"""
+        return self.data.get('sample', sample)['timestamp']
+
+    def _absolute_time_diff(self, time1: int, time2: int) -> int:
+        """Helper to compute how much time has elapsed in _iterate method."""
+        return abs(time1 - time2) / MICROSECONDS_PER_SECOND
+
+    def _iterate(self, starting_annotation: Dict[str, Any], seconds: float, direction: str) -> List[Dict[str, Any]]:
+        """Iterates forwards or backwards in time through the annotations for a given amount of seconds."""
+
+        seconds_with_buffer = seconds + BUFFER
+        starting_time = self._timestamp_for_sample(starting_annotation['sample_token'])
+
+        next_annotation = starting_annotation
+
+        time_elapsed = 0
+
+        annotations = []
+
+        while time_elapsed <= seconds_with_buffer:
+
+            if next_annotation[direction] == '':
+                break
+
+            next_annotation = self.data.get('sample_annotation', next_annotation[direction])
+            current_time = self._timestamp_for_sample(next_annotation['sample_token'])
+
+            time_elapsed = self._absolute_time_diff(current_time, starting_time)
+
+            if time_elapsed < seconds_with_buffer:
+                annotations.append(next_annotation)
+
+        return annotations
+
+    def get_sample_annotation(self, instance: str, sample: str) -> Dict[str, Any]:
+        """Retrives an annotation given an instance token and its sample.
+        :param instance: Instance token
+        :param sample: Sample token
+        """
+        return self.data.get('sample_annotation', self.inst_sample_to_ann[(sample, instance)])
+
+    def get_future_for_agent(self, instance: str, sample: str,
```

Comment:
@freddyaboulton You have a ton of duplicated code between this and `get_past_for_agent`. Can you introduce a function that is called by each but with the appropriate `next` or `prev` flag?

Reply:
This function would be called internally by `get_future_for_agent` and  `get_past_for_agent`.

Reply:
Done!